src/main/java/org/broadinstitute/hellbender/utils/bwa/BwaMemIndex.java

```diff
@@ -30,12 +32,55 @@
  * if that's what you want to do.
  */
 public final class BwaMemIndex implements AutoCloseable {
+
+    /**
+     * Indexing algorithms supported by Bwa.
+     */
+    public enum Algorithm {
+        RB2, IS, AUTO;
+
+        /**
+         * The string name use by the Bwa command line to denote that algorithm.
+         * @return never {@code null} and unique across algorithms.
+         */
+        public String toBwaName() {
+            return this.toString().toLowerCase();
+        }
+    }
+
     private final String indexImageFile; // stash this for error messages
     private volatile long indexAddress; // address where the index was memory-mapped (for use by C code)
     private final AtomicInteger refCount; // keep track of how many threads are actively aligning
     private final List<String> refContigNames; // the reference dictionary from the index
     private static volatile boolean nativeLibLoaded = false; // whether we've loaded the native library or not
 
+    private static final List<String> FASTA_EXTENSIONS = Collections.unmodifiableList(Arrays.asList(".fasta", ".fa"));
+
+
+    public static void indexReference(final String referenceName, final String indexPrefix, final Algorithm algo) {
+        final String referenceFileName = resolveReferenceFileName(referenceName);
+        loadNativeLibrary();
+        createReferenceIndex(referenceFileName, indexPrefix, algo.toBwaName());
+    }
+
+    private static String resolveReferenceFileName(final String referenceName) {
+        if (referenceName == null) {
+            throw new IllegalArgumentException("the reference name provided cannot be null");
+        }
+        final String referenceFileName = FASTA_EXTENSIONS.stream()
+                .filter(ext -> new File(referenceName + ext).isFile()).findFirst().orElse(referenceName);
+        final File referenceFile = new File(referenceFileName);
+        if (referenceFile.isFile()) {
+            if (referenceFile.canRead()) {
+                return referenceFile.getPath();
+            } else {
+                throw new IllegalArgumentException("the reference file provided is not readable: " + referenceFile);
```

Comment:
This is a problem (at least for me) for the whole class and not concretely to this PR: I think that it is important to distinguish when we are trying to index something that it is not a file (or trying to generate the index image for non-existing index) and when a "developer" illegal argument is passed. I would go to throw IO exceptions in that cases to have a way to catch problems with the input files vs. null arguments and other unrelated exceptions. But this will require also to change other parts of the code, so maybe it should be in a different issue.